### PR TITLE
Add support for pod labels and annotations

### PR DESCRIPTION
**Pull request by mjnagel**
Signed-off-by: Micah Nagel <micah.nagel@parsons.com>

#### Summary

This adds two new "options" to the MM spec:

```yaml
spec:
  podTemplate:
    labels: {}
    annotations: {}
```

As outlined in the [issue](https://github.com/mattermost/mattermost-operator/issues/298) the operator handles the precedence of these by always favoring the default labels/annotations, then the user specified (in cases of conflict). I do not believe this changes any current behavior, but happy to discuss if it appears it may.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-operator/issues/298

#### Release Note

```release-note
Add support for labels and annotations to be specified for server pods
```

### CODE REVIEW

**Comment by mjnagel** [comment]
@Szymongib / @gabrieljackson no rush on this, just wanted to make sure you guys got notified that its opened 😄 . Happy to discuss here or in the issue if anything is problematic.

File: pkg/mattermost/mattermost_v1beta.go [hunk 1]
```diff
@@ -220,13 +220,21 @@ func GenerateDeploymentV1Beta(mattermost *mmv1beta.Mattermost, db DatabaseConfig
...
 	if len(mattermost.Spec.LicenseSecret) != 0 {
 		env, vMount, volume, annotations := mattermostLicenceConfig(mattermost.Spec.LicenseSecret)
 		envVarGeneral = append(envVarGeneral, env)
 		volumeMounts = append(volumeMounts, vMount)
 		volumes = append(volumes, volume)
-		podAnnotations = annotations
+		// Add prometheus annotations, overwriting user specified if needed
+		for k, v := range annotations {
+			podAnnotations[k] = v
+		}
```
Comment (Szymongib) [inline pkg/mattermost/mattermost_v1beta.go:237]: Also adding another test for Deployment generation would be much appreciated. On top of the labels and annotations application order, we could verify that the additional `podTemplate.Labels` are applied only to the pods, not all other resources.

**Review by Szymongib — COMMENTED** [review]
Thanks a lot for another contribution @mjnagel! We really appreciate it!

Changes look good, I think we need few small tweaks and some unit tests.

File: apis/mattermost/v1beta1/mattermost_types.go [hunk 2]
```diff
@@ -216,6 +216,14 @@ type PodTemplate struct {
...
 	// +optional
 	ContainerSecurityContext *v1.SecurityContext `json:"containerSecurityContext,omitempty"`
+	// Defines annotations to add to the Mattermost app server pods.
+	// Overrides of default prometheus annotations are ignored.
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+	// Defines labels to add to the Mattermost app server pods.
+	// Overrides what is set in ResourceLabels, does not override default labels (app and cluster labels).
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
```
[the remark this answers is not in the document]
Comment (mjnagel) [reply apis/mattermost/v1beta1/mattermost_types.go:226]: Good idea, I'll rename them that way.

File: apis/mattermost/v1beta1/mattermost_utils.go [hunk 3]
```diff
@@ -201,13 +201,22 @@ func MattermostSelectorLabels(name string) map[string]string {
 // MattermostLabels returns the labels for selecting the resources
 // belonging to the given mattermost.
 func (mm *Mattermost) MattermostLabels(name string) map[string]string {
-	l := MattermostResourceLabels(name)
+	l := map[string]string{}
```
[the remark this answers is not in the document]
Comment (mjnagel) [reply apis/mattermost/v1beta1/mattermost_utils.go:204]: Good catch, I'll add a new function specific for pod labels.

File: apis/mattermost/v1beta1/mattermost_utils.go [hunk 4]
```diff
@@ -201,13 +201,22 @@ func MattermostSelectorLabels(name string) map[string]string {
...
+	// Set resourceLabels ("global") as the initial labels
+	if mm.Spec.ResourceLabels != nil {
+		l = mm.Spec.ResourceLabels
+	}
```
[the remark this answers is not in the document]
Comment (mjnagel) [reply apis/mattermost/v1beta1/mattermost_utils.go:208]: Definitely agreed. I'll try to work on a test and may reach out if I get stuck.

[the remark this answers is not in the document]
Comment (mjnagel) [reply apis/mattermost/v1beta1/mattermost_utils.go:208]: @Szymongib I believe I resolved this and the other comments with https://github.com/mattermost/mattermost-operator/pull/299/commits/7811f53ec7e5f70be5205ab1545b6545a45e0e3d - wasn't sure on the best way to approach the new tests (location/architecture) so let me know if you'd recommend anything different there.

File: pkg/mattermost/mattermost_v1beta.go (hunk 1 above)
Comment (mjnagel) [reply pkg/mattermost/mattermost_v1beta.go:237]: Added - I'm not sure if adding this check the way I did is a good approach or not. Effectively just added a check for the service (no other resources) and placed it in the other generate service tests.

**Review by Szymongib — APPROVED** [review]
LGTM! Thanks!

**Review by gabrieljackson — APPROVED** [review]
Looks good! Thanks for the test coverage.

**Pull request merged**